test/box-luatest/gh_9643_priv_origins_test.lua

```diff
@@ -0,0 +1,182 @@
+        t.assert(not ok)
+        t.assert_equals(err.message, exp_err)
+        t.assert_equals(_priv:get({uid, 'space', sid}), exp)
+
+        -- Check revoke with origins.
+        ok, err = pcall(revoke, 'myuser', 'read', 'space', 'myspace',
+                        {_origin = true})
+        exp_err = "Illegal parameters, options parameter '_origin' should " ..
+            "be of type 'string'"
+        t.assert(not ok)
+        t.assert_equals(err.message, exp_err)
+        t.assert_equals(_priv:get({uid, 'space', sid}), exp)
+
+        ok, err = pcall(revoke, 'myuser', 'write', 'space', 'myspace')
+        exp_err = "User 'myuser' does not have write access on space " ..
+                  "'myspace' provided by default origin"
+        t.assert(not ok)
+        t.assert_equals(err.message, exp_err)
+        t.assert_equals(_priv:get({uid, 'space', sid}), exp)
+
+        ok, err = pcall(revoke, 'myuser', 'read', 'space', 'myspace',
+                        {_origin = 'one'})
+        exp_err = "User 'myuser' does not have read access on space " ..
+                  "'myspace' provided by one origin"
+        t.assert(not ok)
+        t.assert_equals(err.message, exp_err)
+        t.assert_equals(_priv:get({uid, 'space', sid}), exp)
+
+        exp[5] = bit.bor(box.priv.D, box.priv.R)
+        exp[6] = {origins = {
+            [''] = bit.bor(box.priv.D, box.priv.R),
+            one = box.priv.D,
+        }}
+        revoke('myuser', 'write', 'space', 'myspace', {_origin = 'one'})
+        t.assert_equals(_priv:get({uid, 'space', sid}), exp)
+
+        exp[5] = box.priv.D
+        exp[6] = {origins = {
+            [''] = box.priv.D,
+            one = box.priv.D,
+        }}
+        revoke('myuser', 'read', 'space', 'myspace')
+        t.assert_equals(_priv:get({uid, 'space', sid}), exp)
+
+        exp[5] = box.priv.D
+        exp[6] = nil
+        revoke('myuser', 'drop', 'space', 'myspace', {_origin = 'one'})
+        t.assert_equals(_priv:get({uid, 'space', sid}), exp)
+
+        revoke('myuser', 'drop', 'space', 'myspace')
+        t.assert_equals(_priv:get({uid, 'space', sid}), nil)
+    end)
+end
+
+g.test_origins_in_roles = function(cg)
+    cg.server:exec(function()
+        box.schema.role.create('myrole')
+        box.schema.role.create('somerole')
+        local rid = box.space._user.index.name:get('myrole').id
+        local tid = box.space._user.index.name:get('somerole').id
+        local grant = box.schema.role.grant
+        local revoke = box.schema.role.revoke
+        local _priv = box.space._priv
+
+        local exp = {0, rid, 'role', tid, 0}
+        t.assert_equals(_priv:get({rid, 'role', tid}), nil)
+
+        -- Check grant with origins.
+        exp[5] = box.priv.X
+        exp[6] = nil
+        grant('myrole', 'somerole', nil, nil)
+        t.assert_equals(_priv:get({rid, 'role', tid}), exp)
+
+        exp[6] = {origins = {
+            [''] = box.priv.X,
+            two = box.priv.X,
+        }}
```

Comment:
Better write the test so that it doesn't break if we add new priv options.

Reply:
True, fixed.